docs/design/event_compression.md

```diff
@@ -0,0 +1,79 @@
+# Kubernetes Event Compression
+
+This document captures the design of event compression.
+
+
+## Background
+
+Kubernetes components can get into a state where they generate tons of events which are identical except for the timestamp. For example, when pulling a non-existing image, Kubelet will repeatedly generate ```image_not_existing``` and ```container_is_waiting``` events until upstream components correct the image. When this happens, the spam from the repeated events makes the entire event mechanism useless. It also appears to cause memory pressure in etcd (see [#3853](https://github.com/GoogleCloudPlatform/kubernetes/issues/3853)).
+
+## Proposal
+Each binary that generates events (for example, ```kubelet```) should keep track of previously generated events so that it can collapse recurring events into a single event instead of creating a new instance for each new event.
+
+Event compression should be best effort (not guaranteed). Meaning, in the worst case, ```n``` identical (minus timestamp) events may still result in ```n``` event entries.
+
+## Design
+Instead of a single Timestamp, each event object [contains](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/types.go#L1111) the following fields:
+ * ```FirstTimestamp util.Time``` 
+   * The date/time of the first occurrence of the event.
+ * ```LastTimestamp util.Time```
+   * The date/time of the most recent occurrence of the event.
+   * On first occurrence, this is equal to the FirstTimestamp.
+ * ```Count int```
+   * The number of occurrences of this event between FirstTimestamp and LastTimestamp
+   * On first occurrence, this is 1.
+
+Each binary that generates events will:
+ * Maintain a new global hash table to keep track of previously generated events (see ```pkg/client/record/events_cache.go```).
+ * The code that “records/writes” events (see ```StartRecording``` in ```pkg/client/record/event.go```), uses the global hash table to check if any new event has been seen previously.
+   * The key for the hash table is generated from the event object minus timestamps/count/transient fields (see ```pkg/client/record/events_cache.go```), specifically the following events fields are used to construct a unique key for an event:
+     * ```event.Source.Component```
+     * ```event.Source.Host```
+     * ```event.InvolvedObject.Kind```
+     * ```event.InvolvedObject.Namespace```
+     * ```event.InvolvedObject.Name```
+     * ```event.InvolvedObject.UID```
+     * ```event.InvolvedObject.APIVersion```
+     * ```event.Reason```
```

Comment:
Actually, this is not hypothetical.  I think it can easily happen today.
If a pod is repeatedly terminating and being restarted by an RC, and you have a lot of nodes, then you are going to see this Message:
```Successfully assigned kibana-logging-controller-$FOO to kubernetes-minion-$BAR.c.saad-dev-vms.internal```
a bunch of times with different values of $FOO and $BAR.
So, this needs attention sooner rather than later.